kernel/arch/riscv64/Interrupts.cpp

```diff
@@ -39,73 +40,93 @@ extern "C"
         "ecall from U-mode",
         "ecall from S-mode"
     };
+    constexpr size_t ExceptionStrCount = sizeof(ExceptionStrs) / sizeof(const char*);
+    constexpr uintptr_t InterruptBitMask = 1ul << 63;
+
+    void HandleException(Npk::TrapFrame* frame)
+    {
+        using namespace Npk;
+        using namespace Npk::Memory;
+
+        Memory::VmFaultFlags faultFlags {};
+        switch (frame->vector)
+        {
+        case 12:
+            faultFlags |= VmFaultFlag::Execute;
+            break;
+        case 13:
+            faultFlags |= VmFaultFlag::Read;
+            break;
+        case 14:
+            faultFlags |= VmFaultFlag::Write;
+            break;
+        //TOOD: UD and decoder for FPU/vector register traps
+
+        default:
+            {
+                const char* exceptionName = frame->vector < ExceptionStrCount 
+                    ? ExceptionStrs[frame->vector] : "unknown";
+                Log("Unexpected exception: %s (%lu) @ 0x%lx, sp=0x%lx, ec=0x%lx", LogLevel::Fatal,
+                   exceptionName, frame->vector, frame->sepc, frame->sp, frame->ec);
+            }
+        };
+
+        ASSERT_(faultFlags.Any());
+        if (frame->ec < hhdmBase)
+            ASSERT(VMM::CurrentActive(), "nullptr deref in kernel?");
+        if (frame->flags.spp == 0)
+            faultFlags |= VmFaultFlag::User;
+
+        bool success = false;
+        if (frame->ec < hhdmBase)
+        {
+            ASSERT(VMM::CurrentActive(), "nullptr deref in kernel?");
+            success = VMM::Current().HandleFault(frame->ec, faultFlags);
+        }
+        else
+            success = VMM::Kernel().HandleFault(frame->ec, faultFlags);
+        if (!success)
+            Log("Bad page fault @ 0x%lx, flags=0x%lx", LogLevel::Fatal, frame->ec, faultFlags.Raw());
+    }
     
     void TrapDispatch(Npk::TrapFrame* frame)
     {
         using namespace Npk;
-        const RunLevel prevRunLevel = CoreLocal().runLevel;
-        CoreLocal().runLevel = RunLevel::IntHandler;
-        Tasking::Scheduler::Global().SavePrevFrame(frame, prevRunLevel);
+        using namespace Npk::Tasking;
+
+        const RunLevel prevRl = RaiseRunLevel(RunLevel::Interrupt);
+        if (prevRl == RunLevel::Normal)
+            *ProgramManager::Global().GetCurrentFrameStore() = frame;
+        EnableInterrupts();
 
-        const bool isInterrupt = frame->vector & (1ul << 63);
-        frame->vector &= ~(1ul << 63);
+        const bool isInterrupt = frame->vector & InterruptBitMask;
+        frame->vector &= ~InterruptBitMask;
 
         if (isInterrupt)
         {
             switch (frame->vector)
             {
-            case 1:
+            case 1: //IPI
```

Comment:
referring to the previous comment, i think this way too could work, if we don't want to add define for everything.